transport/tchannel/handler.go

```diff
@@ -86,31 +88,39 @@ func (h handler) Handle(ctx ncontext.Context, call *tchannel.InboundCall) {
 }
 
 func (h handler) handle(ctx context.Context, call inboundCall) {
-	start := time.Now()
-	err := h.callHandler(ctx, call, start)
-	if err == nil {
-		return
+	// you MUST close the responseWriter no matter what
+	responseWriter := newResponseWriter(call.Response(), call.Format())
+
+	handlerErr := h.callHandler(ctx, call, responseWriter)
+	if handlerErr != nil {
+		// we have an error, so we're going to propagate it as a yarpc error,
+		// regardless of whether or not it is a system error.
+		yarpcError := yarpcerrors.ToYARPCError(handlerErr)
+		// TODO: what to do with error? we could have a whole complicated scheme to
+		// return a SystemError here, might want to do that
+		text, _ := yarpcerrors.ErrorCode(yarpcError).MarshalText()
+		responseWriter.addHeader(ErrorCodeHeaderKey, string(text))
+		if name := yarpcerrors.ErrorName(yarpcError); name != "" {
+			responseWriter.addHeader(ErrorNameHeaderKey, name)
+		}
+		if message := yarpcerrors.ErrorMessage(yarpcError); message != "" {
+			responseWriter.addHeader(ErrorMessageHeaderKey, message)
+		}
 	}
-
-	if _, ok := err.(tchannel.SystemError); ok {
+	systemError, ok := getSystemError(handlerErr)
+	if err := responseWriter.Close(ok); err != nil {
 		// TODO: log error
-		_ = call.Response().SendSystemError(err)
+		_ = call.Response().SendSystemError(tchannel.NewSystemError(tchannel.ErrCodeUnexpected, err.Error()))
```

Comment:
we've already passed the call response into the responsewriter, we shouldn't call into it directly here.